### ISSUE TITLE
Inquiry about PGD adversarial image generation code

### ISSUE THREAD

**Issue by xiyuzzz-njust** [ASTRAL-Group/ASTRA#1]
I have been studying your recent paper, "Steering Away from Harm: An Adaptive Approach to Defending Vision Language Model Against Jailbreaks," with great interest. The adaptive steering approach you propose is very compelling.

Are the adversarial images in the code repository compatible for direct use with the LLaVA-v1.5-7B model for image attribution?

Would you be willing to disclose the specific code you use to generate PGD adversarial images?  I would like to replicate your excellent work on more models.

**Comment by rookiehb** [comment]
No, you should generate adversarial images specifically for the LLaVA-v1.5-7B. 

You can follow this repo (https://github.com/Unispac/Visual-Adversarial-Examples-Jailbreak-Large-Language-Models) to help you generate some adversarial examples.

**Comment by xiyuzzz-njust** [comment]
Can the design of hyperparameters for generating adversarial images corresponding to the LLaVA-v1.5-7b model using the PGD method refer to the settings in the paper? 

Will these hyperparameters have an impact on defense effectiveness?

**Comment by rookiehb** [comment]
I don't think that can make a huge difference, as long as the loss converges when doing PGD attacks.

**Comment by xiyuzzz-njust** [comment]
May I ask about the idea of PGD attack, which is to obtain all targets from advbench and anthropoic-hhh as optimization targets, and then generate 16 adversarial images based on the code in this repo (https://github.com/Unispac/Visual-Adversarial-Examples-Jailbreak-Large-Language-Models) ?

This process does not require inputting malicious questions, and ultimately generates a universal jailbreak image, right?

**Comment by rookiehb** [comment]
Yes, when generating adversarial images for training, we do not use malicious questions as input during the PGD attacks.

During evaluation, however, adversarial images can be generated either with or without malicious instructions as input. In our main experiments, we evaluate using universal jailbreak images, which are generated without specific malicious instructions. In our Jailbreak transferability experiments, we use adversarial images crafted with both malicious instructions and affirmation responses, which just need some simple modification based on the existing repo.

**Comment by xiyuzzz-njust** [comment]
I calculated the steering vector and calibration activation based on the description in your paper. Subsequently, I used the code from your repository to add steering vectors to the 14 layers of the LLaVA-v1.5-7b model. 

At first, the alpha value was set to 10, but I found that there were a large number of repetitions in the model's response. Then, when the alpha value was set to 5, I found that there were still a large number of repetitions in the model's response. Is this normal?

**Comment by rookiehb** [comment]
No, if you check our steering vectors, it won't harm the consistency of the response. Maybe you can print the norm of the hidden states and steering vectors to see why this happens.

**Comment by xiyuzzz-njust** [comment]
I found that during the inference process of the LLaVA-v1.5-13b model, the calculated vector angle clip_proj is approximately 0.04, and sometimes it is negative and can be changed to 0 through torch.clamp

However, during the inference process of the LLaVA-v1.5-7b model, the calculated vector angle clip_proj is approximately 0.2, and the fluctuation is very small and will not be negative, which may lead to oversteering. I think this is a possible reason. Because I set the alpha value to 3 later on, the presence of a large number of repetitions in the model's response will almost disappear, but it seems that the defense effect is not very significant.

Is this related to the selected images from ImageNet?  Would you be willing to publicly release the code for building standard activation?

**Comment by rookiehb** [comment]
Could you share how you reproduce the method in each step in detail? Also, which sanity checks did you do to ensure each step is correct?

**Comment by xiyuzzz-njust** [comment]
Firstly, I randomly selected 16 images of different categories from ImageNet, and then I apply PGD attack for 2500 iterations  with a step size of 1/1020. And I checked that the loss had converged. Subsequently, the code in the repository is used to extract the steering vectors for each layer of the model. The model used is liuhaotian/llava-v1.5.

When I execute this file (https://github.com/ASTRAL-Group/ASTRA/blob/main/extract_attr/extract_llava_jb_attr.py), I only get 43 queries and 43 data_adv_img. The model used is llava-hf/llava-1.5.

To extract calibration activation, I first randomly selected 21 images of different categories from Imagenet, and then applied a forward to obtain the hidden layer activation value of each layer.
[extracting_activations_llava_benign.txt](https://github.com/user-attachments/files/20449044/extracting_activations_llava_benign.txt)

For the LLaVA-1.5-13b model, the norm of the reference'activations provided by your repository is 116.0, while the norm of the reference'activations calculated by me is 63.84375, resulting in a significant difference in the calculated clip_proj. 

And I found that the norm of the activation at level 20 during inference of the model is in the range of 50 to 70.

**Comment by rookiehb** [comment]
For the adv-images, have you checked the ASR w/o defense? It seems that the jailbreak information in the adv_images isn't enough, cuz you could only get 43 pairs. It would be better if you could share the code you generated adv_images, the loss plot, ASR w/o defense, and responses from the model. Also, you can uncomment line 49 in this file (https://github.com/ASTRAL-Group/ASTRA/blob/main/extract_attr/extract_llava_jb_attr.py) and email the log to me as well. 

For the reference activation, we average all the activations of generated tokens, instead of the activation in the last input token.

**Comment by tunglamlqddb** [comment]
Hi,
I'm sorry to bring this up.
I am also trying to generate adv images for Qwen2.5-VL-7B, but I'm stuck at passing the adv_image through the [processor of qwen-vl](https://github.com/huggingface/transformers/blob/f4fc42216cd56ab6b68270bf80d811614d8d59e4/src/transformers/models/qwen2_vl/image_processing_qwen2_vl.py#L235), which converts adv_image from tensor to numpy array, hence breaking the gradient of the loss w.r.t the noise.

Could you share the code you used to generate adv_imgs for this model?
Thank you in advance.

**Comment by rookiehb** [comment]
Sure, send me an email and I will share the code with you.

**Comment by tunglamlqddb** [comment]
Thank you @rookiehb  
Here is my gmail: tunglamlqddb@gmail.com

**Comment by xiyuzzz-njust** [comment]
I'm sorry to disturb you again.
I am also curious about how to generate adv_img for Qwen2-VL-7B, so would you be willing to share your code with me?
My gmail address is: xiyuceng76@gmail.com
I will be very grateful to you.

**Comment by stha-prashant** [comment]
Hi @rookiehb, could you also provide me the code for generating adv_img for Qwen2-VL-7B (prashantshrestha328@gmail.com). Additionally, would it be possible to provide the `./results/adv_img_attr/` folder of adversarial images used for calculating the steering vector in the paper? Thanks

**Comment by rookiehb** [comment]
Hi, I have sent you the link to the code. For ./results/adv_img_attr/, I think you can follow the demos to get these images.
